### PR TITLE
Share the etcd and redis watcher between sessions

### LINKED ISSUE #715: Session: share the etcd watch stream between different sessions to avoid duplicated metadata updates
Describe your problem
---------------------

as titled.

### PR DESCRIPTION
**Pull request by shihao-thx**
Signed-off-by: shihao-thx <shihao@zju.edu.cn>

What do these changes do?
-------------------------

as titled.

Related issue number
--------------------

Fixes #715.

### CODE REVIEW

**Comment by shihao-thx** [comment]
Hi, @sighingnow. Sharing the etcd watcher has already done. If there are any problems, please let me know. Thanks!

**Comment by sighingnow** [comment]
> Hi, @sighingnow. Sharing the etcd watcher has already done. If there are any problems, please let me know. Thanks!

Will take a review later today.

File: src/server/services/etcd_meta_service.cc [hunk 1]
```diff
@@ -390,4 +397,30 @@ Status EtcdMetaService::preStart() {
...
+    std::string sessionID, const std::shared_ptr<EtcdMetaService>& pems,
+    asio::io_context& ctx,
+    callback_t<const std::vector<IMetaService::op_t>&, unsigned,
+               callback_t<unsigned>>
+        callback,
+    std::string const& prefix, unsigned since_rev,
+    std::unique_ptr<etcd::Client>& etcd_) {
+  std::lock_guard<std::mutex> scope_lock(session_add_mutex_);
+  sessions_.emplace_back(
+      std::make_shared<SessionInfo>(sessionID, pems, ctx, callback, prefix));
```
Comment (sighingnow) [inline src/server/services/etcd_meta_service.cc:410]: Use a `std::map<SessionID, SessionInfo>` to simplify logic in handlers.

File: src/server/services/etcd_meta_service.cc [hunk 2]
```diff
@@ -52,37 +54,47 @@ void EtcdWatchHandler::operator()(etcd::Response const& resp) {
...
+            // invalid event type.
+            break;
+          }
+          }
+          break;
+        }
+        ++it;
+      } else {
+        it = sessions_.erase(it);
+      }
```
Comment (sighingnow) [inline src/server/services/etcd_meta_service.cc:97]: I suggest you first split and group events to sessions, then dispatch events to each session, e.g.,

```
std::map<std::string /*prefix*/, std::vector<etcd::Event>>  event_groups;

for (auto const &event: resp.events()) {
    std::string const& key = event.kv().key();
    if (key.starts_with("/s")) { // it should belong to some session, and you know the length of session prefix
        event_groups[key.substr(1, 8 /* maybe */)].emplace_back(event);
    }
}

// then dispatch events to sessions

for (auto const &group: event_groups) {
    auto const &session = sessions_[group.first];
    for (auto const &event: group.second) {
         // processing the events, as existing logic
    }
}

```

File: src/server/services/etcd_meta_service.h [hunk 3]
```diff
@@ -61,38 +61,58 @@ class EtcdMetaService;
...
+  }
+
+  std::shared_ptr<etcd::Watcher> getWatcher(
+      std::string sessionID, const std::shared_ptr<EtcdMetaService>& pems,
+      asio::io_context& ctx,
+      callback_t<const std::vector<IMetaService::op_t>&, unsigned,
+                 callback_t<unsigned>>
+          callback,
+      const std::string& prefix, unsigned since_rev,
+      std::unique_ptr<etcd::Client>& etcd_);
```
Comment (sighingnow) [inline src/server/services/etcd_meta_service.h:97]: Looks strange that `WatcherHanlder` has method to initialize a `Watcher` ....

**Review by sighingnow — CHANGES_REQUESTED** [review]
Please implement the `EtcdWatchHandler::operator()` as review comments.

**Comment by sighingnow** [comment]
If it is possible, I suggest you consider unify the watcher handler logic in both etcd metadata service and the redis metadata service, as they share a lot of common stuffs.

**Comment by shihao-thx** [comment]
Thanks for the suggestions. I'll do that as soon as possible.

**Comment by shihao-thx** [comment]
I modified the way to dispatch watch events to sessions as suggested. If it's fine, I'll try to apply it to sharing redis metadata watcher in a few days.

**Comment by sighingnow** [comment]
> I modified the way to dispatch watch events to sessions as suggested. If it's fine, I'll try to apply it to sharing redis metadata watcher in a few days.

Please go ahead.

**Comment by shihao-thx** [comment]
I noticed the `spill_test` failed, but it can pass on my PC. Should I deal with it?

**Comment by sighingnow** [comment]
The crash happened inside vineyardd:

image

Looks caused by this pull request.

**Comment by shihao-thx** [comment]
Hi, @sighingnow . Sharing the redis watcher has already done too. If there are any problems, please let me know. Thanks!

**Review by sighingnow — APPROVED** [review]
LGTM now.

Will merge after repeat the CI for several times to confirm the correctness.

**Comment by sighingnow** [comment]
Merged. Thanks!

**Pull request merged**